src/sentry/rules/history/preview.py

```diff
@@ -198,34 +199,66 @@ def get_fired_groups(
 
 
 def get_top_groups(
-    project: Project, start: datetime, end: datetime, condition_activity: GroupActivityMap
+    project: Project,
+    start: datetime,
+    end: datetime,
+    condition_activity: GroupActivityMap,
+    dataset_map: Dict[int, Dataset],
 ) -> GroupActivityMap:
     """
     Filters the activity to contain only groups that have the most events in the past 2 weeks.
 
     Since frequency conditions require one snuba query per groups, we need to limit the number groups we process.
     """
-    if not condition_activity:
-        return condition_activity
-    # TODO: Also check other datasets for top groups
-    groups = raw_query(
-        dataset=Dataset.Events,
-        start=start,
-        end=end,
-        filter_keys={"project_id": [project.id]},
-        conditions=[("group_id", "IN", list(condition_activity.keys()))],
-        aggregations=[("count", "group_id", "groupCount")],
-        groupby=["group_id"],
-        order_by="-groupCount",
-        seleted_columns=["group_id"],
-        limit=FREQUENCY_CONDITION_GROUP_LIMIT,
-    ).get("data", [])
-
-    top_groups = {group["group_id"] for group in groups}
-    top_activity = {
+    datasets = {dataset_map.get(group) for group in condition_activity.keys()}
+
+    # queries each dataset for top x groups and then gets top x overall
+    query_params = []
+    for dataset in datasets:
+        if dataset is None:
+            continue
+        group_ids = list(condition_activity.keys())
+        kwargs = {
+            "dataset": dataset,
+            "start": start,
+            "end": end,
+            "filter_keys": {"project_id": [project.id]},
+            "conditions": [("group_id", "IN", group_ids)],
+            "aggregations": [("count", "group_id", "groupCount")],
+            "groupby": ["group_id"],
+            "order_by": "-groupCount",
+            "selected_columns": ["group_id", "groupCount"],
+            "limit": FREQUENCY_CONDITION_GROUP_LIMIT,
+        }
+        if dataset == Dataset.Transactions:
+            kwargs["conditions"] = [[["hasAny", ["group_ids", ["array", group_ids]]], "=", 1]]
+            kwargs["aggregations"].append(("arrayJoin", ["group_ids"], "group_id"))
+        query_params.append(SnubaQueryParams(**kwargs))
+
+    groups = []
+    for result in bulk_raw_query(query_params):
+        groups.extend(result.get("data", []))
+
+    sorted_groups = sorted(groups, key=lambda x: int(x["groupCount"]), reverse=True)
+
+    top_groups = {group["group_id"] for group in sorted_groups[:FREQUENCY_CONDITION_GROUP_LIMIT]}
```

Comment:
make this a `set()`

Reply:
i think this is a set

Reply:
1. you're right
2. not if it's empty lol
https://stackoverflow.com/a/17373237

Reply:
sorry i meant I didn't realize that but I think the point still holds in case we get empty results back. could add a test for `no preview results` to confirm

Reply:
ah you're right. no problem then

Reply:
lol yeah i'll add a test for that